operators/src/pro/cache/tile_cache.rs

```diff
@@ -29,16 +30,425 @@ pub struct TileCacheBackend {
+        let old_cache_entry = cache.insert(cache_entry_id, cache_entry);
+        let old_lru_entry = cache_lru.push(cache_entry_id, key.clone());
+        assert!(old_lru_entry.is_none()); // this should always work, because we just inserted a new CacheEntryId
+        assert!(old_cache_entry.is_none()); // this should always work, because we just inserted a new CacheEntryId
+
+        // cache bound can be temporarily exceeded as the entry is moved form the landing zone into the cache
+        // but the total of cache + landing zone is still below the bound
+        cache_size.add_bytes_allow_overflow(cache_entry_size_bytes);
+
+        debug!(
+            "Finished query {}. Cache size: {}. Cache size used: {}, Cache used percentage: {}.",
+            query_id,
+            cache_size.total_byte_size(),
+            cache_size.byte_size_used(),
+            cache_size.size_used_fraction()
+        );
+
+        // We now evict elements from the cache until bound is satisfied again
+        self.evict_elements_until_cache_size_is_satisfied();
+
+        Ok(())
+    }
+
+    fn evict_elements_until_cache_size_is_satisfied(&mut self) {
+        while self.cache_size.is_overfull() {
+            // this should always work, because otherwise it would mean the cache is not empty but the lru is.
+            // the landing zone is smaller than the cache size and the entry must fit into the landing zone.
+            if let Some((pop_entry_id, pop_entry_key)) = self.lru.pop_lru() {
+                let (cache, cache_size, _cache_lru) = self
+                    .cache_and_size_and_lru_mut(&pop_entry_key)
+                    .expect("There must be cache elements entry if there was an LRU entry");
+                {
+                    let old_cache_entry = cache
+                        .remove(&pop_entry_id)
+                        .expect("LRU entry must be in cache");
+                    cache_size.remove_element_bytes(&pop_entry_id);
+                    cache_size.remove_element_bytes(&old_cache_entry);
+                }
+
+                debug!(
+                    "Evicted query {}. Cache size: {}. Cache size used: {}, Cache used percentage: {}.",
+                    pop_entry_id,
+                    cache_size.total_byte_size(),
+                    cache_size.byte_size_used(),
+                    cache_size.size_used_fraction()
+                );
+            } else {
+                panic!("Cache is overfull but LRU is empty. This must not happen.");
+            }
+        }
+    }
+
+    fn remove_querys_from_cache_and_lru(
+        &mut self,
+        key: &CanonicOperatorName,
+        cache_entry_ids: &[CacheEntryId],
+    ) {
+        if let Some((cache, cache_size, cache_lru)) = self.cache_and_size_and_lru_mut(key) {
+            for cache_entry_id in cache_entry_ids {
+                let _old_lru_entry = cache_lru.pop_entry(cache_entry_id);
+                if let Some(old_cache_entry) = cache.remove(cache_entry_id) {
```

Comment:
it's nice that you wrapped the functionality for calculating the cache size. Would it make sense to wrap the whole cache Hashmap and the lru type with a custom type that autoamtically adjusts the byte size on remove/pop so that we can't forget it?

Reply:
yes, will try to do that when i finally start building the vector cache